capsules/src/sht3x.rs

```diff
@@ -0,0 +1,308 @@
+    Idle,
+    ReadStatus,
+    Read,
+    ReadData,
+    Reset,
+}
+
+fn crc8(data: &[u8]) -> u8 {
+    let polynomial = 0x31;
+    let mut crc = 0xff;
+
+    for x in 0..data.len() {
+        crc ^= data[x as usize] as u8;
+        for _i in 0..8 {
+            if (crc & 0x80) != 0 {
+                crc = crc << 1 ^ polynomial;
+            } else {
+                crc = crc << 1;
+            }
+        }
+    }
+    crc
+}
+
+pub struct SHT3x<'a, A: Alarm<'a>> {
+    i2c: &'a dyn i2c::I2CDevice,
+    humidity_client: OptionalCell<&'a dyn kernel::hil::sensors::HumidityClient>,
+    temperature_client: OptionalCell<&'a dyn kernel::hil::sensors::TemperatureClient>,
+    state: Cell<State>,
+    buffer: TakeCell<'static, [u8]>,
+    read_temp: Cell<bool>,
+    read_hum: Cell<bool>,
+    alarm: &'a A,
+}
+
+impl<'a, A: Alarm<'a>> SHT3x<'a, A> {
+    pub fn new(
+        i2c: &'a dyn i2c::I2CDevice,
+        buffer: &'static mut [u8],
+        alarm: &'a A,
+    ) -> SHT3x<'a, A> {
+        SHT3x {
+            i2c: i2c,
+            humidity_client: OptionalCell::empty(),
+            temperature_client: OptionalCell::empty(),
+            state: Cell::new(State::Idle),
+            buffer: TakeCell::new(buffer),
+            read_temp: Cell::new(false),
+            read_hum: Cell::new(false),
+            alarm: alarm,
+        }
+    }
+
+    pub fn clear_status(&self) {
+        self.buffer.take().map(|buffer| {
+            self.i2c.enable();
+
+            buffer[0] = ((Registers::CLEARSTATUS as u16) >> 8) as u8;
+            buffer[1] = Registers::CLEARSTATUS as u8;
+            self.i2c.write(buffer, 2);
+
+            self.state.set(State::Reset);
+        });
+    }
+
+    pub fn read_status(&self) {
+        self.buffer.take().map(|buffer| {
+            self.i2c.enable();
+
+            buffer[0] = Registers::READSTATUS as u8;
+            self.i2c.write_read(buffer, 1, 1);
+
+            self.state.set(State::ReadStatus);
+        });
+    }
+
+    pub fn reset(&self) {
+        self.buffer.take().map(|buffer| {
+            self.i2c.enable();
+
+            buffer[0] = ((Registers::SOFTRESET as u16) >> 8) as u8;
+            buffer[1] = Registers::SOFTRESET as u8;
+            self.i2c.write_read(buffer, 2, 2);
+            self.state.set(State::Reset);
+        });
+    }
+
+    pub fn heater(&self, h: bool) {
+        self.buffer.take().map(|buffer| {
+            self.i2c.enable();
+
+            if h == true {
+                buffer[0] = ((Registers::HEATEREN as u16) >> 8) as u8;
+                buffer[1] = Registers::HEATEREN as u8;
+            } else {
+                buffer[0] = ((Registers::HEATERDIS as u16) >> 8) as u8;
+                buffer[1] = Registers::HEATERDIS as u8;
+            }
+
+            self.i2c.write(buffer, 2);
+
+            self.state.set(State::Reset);
+        });
+    }
+
+    pub fn is_heater_enabled(&self) -> bool {
+        false
+    }
+
+    pub fn read_humidity(&self) -> ReturnCode {
```

Comment:
Should these functions be public? I'm guessing one is supposed to access them through the HILs?

Reply:
You are right, I updated the code so the functions are not public anymore